GuidedTours: change highlighting to a border and an arrow

lsinger opened the pull request:
We have gotten feedback that it's not always very clear which element a tour step is pointing towards.  Additionally the contrast between the step itself and the rest of Calypso wasn't high enough to make it stand out. This PR adds a glow around steps and an arrow that points towards steps' targets. This should  make clearer that there is a step and where it is referring to.

This new design has four goals:

- increase the visibility and contrast of the steps themselves to make clear that they are not really a part of the functional interface but a separate meta thing
- make clearer to the user what target element they are expected to look at, as previous solutions (see below) were too subtle based on what we heard from testers
- couple the technical implementation to the target element that the step is pointing at, as previous attempts eg with fading out everything but the target element proved too unreliable when taking scrolling and resizing into account
- replace previous attempts at fading out the rest of Calypso, since that removed any points of reference the user might have had in the UI

Previously, with pulsating dot:

![pulsating-dot](https://cloud.githubusercontent.com/assets/23619/15722548/8aced2ee-283e-11e6-9664-248d0fcae5ba.gif)

Previously, with fading out the rest of Calypso:

![fade-out](https://cloud.githubusercontent.com/assets/23619/15722544/8217f6bc-283e-11e6-8827-da83ddcc0d22.gif)

The previous method of highlighting target elements by fading out the rest of Calypso turned out to be not reliable enough with scrolling and resizing taken into account. As an alternative, we are now trying a method that decorates the actual target, and should thus be less fragile. 

GIF of the updated design proposed in this PR:

![arrow-highlights](https://cloud.githubusercontent.com/assets/23619/15712544/f74f1482-2812-11e6-8b44-16ec88b9683f.gif)

To test:

- open http://calypso.localhost:3000/?tour=main and go through the tour
- make sure the glowing border and arrow look correct, and that the arrow points to a reasonable element
- test this works on different screen sizes as well, i.e. desktop / tablet / phone

Props @ehg @shaunandrews @mcsf

folletto commented:
Looks good!
Tested on Chrome, Safari, Firefox: 👍 

Aside, quick Calypso live branch link to test it → https://calypso.live/?tour=main&branch=add/guided-tours-highlight-arrow

lsinger commented:
> would be worth to add the same border to everything Tours? Including the first and the last boxes?

What does @shaunandrews think? 

Also cc @shaunandrews on this in general. :)

drw158 commented:
Hm, I don't think this new style looks very good. The big blue border feels like it's forcing itself to grab the viewer's attention. The blue border is very strong. It should at least have the same style as our focus input elements, if you want to use that style. Imo, it would be much less messy just to use one ~2px border instead of two. Here are some alternative solutions:

- I would try using the our orange color in some way. Maybe in the text or buttons? An orange border?
- Icons: I believe an icon with some color would attract the viewer's eye.
- Animation: a subtle, gentle animation would help. Maybe a swaying motion? The original pulsing dot that I've seen us use would also work. That would solve 2 problems: points to an element and grabs attention.

I believe fading out the rest of the UI is probably the best option to highlight the tour step. If that is impossible, then maybe we could go in a different direction. We could contain tours within a stepped modal, and show pieces of the UI within the modal, instead of showing a tooltip on the UI itself.

Also, could you update all of the colors to use variables?

![](http://cdn.pttrns.com/536/6287_f.jpg)

folletto commented:
> I believe fading out the rest of the UI is probably the best option to highlight the tour step. If that is impossible, then maybe we could go in a different direction. We could contain tours within a stepped modal, and show pieces of the UI within the modal, instead of showing a tooltip on the UI itself.

What I think is the strong and excellent idea Tours are following is to avoid doing either different UIs or obscuring content. This means that while sure, the dialog might be there, I can just go on with what I'm doing and ignore it, and this happens at every step.

It's something borrowed from game design, where the "tutorial" is done inside the actual, real UI, and you can ignore it as long as you want, or pick it up again when you want. This is more effective because we can create smaller Tours that never yank the user away from the actual, real, interface they'll be using every day. It's _immersive_ instead of disruptive. :)

And I think that's powerful and worth pursuing. :)

shaunandrews commented:
> would be worth to add the same border to everything Tours? Including the first and the last boxes?

I think the first box (which draws attention to itself using an entrance animation) is more of an introduction—not part of the tour— and doesn't need the border.

The last box, however, is really a part of the tour, and should have the same treatment as the other steps.

> The big blue border feels like it's forcing itself to grab the viewer's attention.

Thats the intention.

> It should at least have the same style as our focus input elements

I was trying to mimic those styles in my mockups, but with a slightly bigger blue border/shadow for more emphasis.

> The nudge component uses some of these ideas to create something attention grabbing, but not garish

The nudge has a totally different goal; Its an upsell. Nudges are supposed to be subtly noticeable. The tour is meant to be in your face. Perhaps garishly noticeable.

> We could contain tours within a stepped modal, and show pieces of the UI within the modal, instead of showing a tooltip on the UI itself.

I think its really important to introduce the UI in its natural habitat. Presenting it as separate pieces outside of their normal context could really reduce the effectiveness of the tour.

lsinger commented:
@drw158 thanks so much for your input! I'm sorry that I wasn't clearer in the PR with regard to what we are trying to achieve. I've updated the PR description accordingly. 

I hope it now explains better why fading out the UI, modals, or using colors that make the steps blend in with the UI would conflict with our goals. I've also updated the colors to use variables ✅.

lsinger commented:
> The last box, however, is really a part of the tour, and should have the same treatment as the other steps.

Updated—also moved the final step down a bit.

drw158 commented:
Thank you for putting up before screenshots.

I think the before screenshots are much much better in terms of style and UI/UX. Even without the fading I think it's very nice. The new design is disappointing to me because it feels like a retrogression in my opinion.

It would be nice to understand what issues the fading was causing and if we can address and fix those issues.

I think just eliminating the fading would be more than enough. The pulsing dots are nice, and I have a feeling it would perform well in user tests.

> The big blue border feels like it's forcing itself to grab the viewer's attention.
> 
> Thats the intention.

Yes, I understand that the intention is to grab the viewer's attention. The _way it is doing it_ is what I'm not a fan of. I feel like the new design is yelling at me, instead of grabbing my hand and guiding me.

> The nudge has a totally different goal; Its an upsell. Nudges are supposed to be subtly noticeable. The tour is meant to be in your face. Perhaps garishly noticeable.

I'm not suggesting you copy the design of the nudges. But I think the ideas could be incorporated. Color, icon, etc.

> I think its really important to introduce the UI in its natural habitat. Presenting it as separate pieces outside of their normal context could really reduce the effectiveness of the tour.

I totally agree, and I think the modal idea is sort a last resort. It is an alternate solution if the contextual tips are not succeeding.

lsinger commented:
Based on a screenshot from @shaunandrews I've reworked things to a darker style. A bunch of colors probably still need adjusting, please let me know (or experiement yourself) what colors we should try where. E.g. the blue link on the last steps seems too dark, and similarly the "Click the W to continue" seems too dark for the background (which currently is `darken( $gray, 52% )`). 

GIF of the whole tour:

![dark](https://cloud.githubusercontent.com/assets/23619/15773271/58a2efa6-2976-11e6-9f5c-6375ef907a78.gif)

/cc @drw158 @shaunandrews @folletto

shaunandrews commented:
Tweaked the colors to fit with the new gray-dark-vader colors, and added some transitions for each step to help draw you eye.

folletto commented:
The new color scheme seems working well for me. Nice job. :)

Tested on: Chrome, Firefox, Safari, Edge 👍  
Seems not working in IE11: the first popup shows, but clicking it doesn't seem to work. Console shows:

![screen shot 2016-06-03 at 16 32 44](https://cloud.githubusercontent.com/assets/4389/15784081/d2c51304-29a8-11e6-9994-81a6d4df61a2.png)


Two bits I noticed:

1. The tour doesn't dismiss if the user ignores it at some point and just clicks around somewhere else. It feels to me in that scenario the tour should stop itself, and let the user to get around freely. 
2. When clicked on "Learn more" in the final dialog the dialog didn't close once I got back to it. I felt it should have closed, but that might be by choice. I felt it was worth mentioning.

Neither is part of this PR tho, so let me know if it's worth opening separate issues for these.

shaunandrews commented:
> The tour doesn't dismiss if the user ignores it at some point and just clicks around somewhere else. It feels to me in that scenario the tour should stop itself, and let the user to get around freely.

I had the same experience a few times. It really easy to click somewhere else and still have the tour open — and it just doesn't make sense. Worth an issue/PR.

> When clicked on "Learn more" in the final dialog the dialog didn't close once I got back to it. I felt it should have closed, but that might be by choice. I felt it was worth mentioning.

I don't have a strong opinion either way. But worth an issue.

folletto commented:
:y: done → #5787 and #5788 

Thanks!

coreh commented on client/layout/guided-tours/config.js:
```diff
@@ -51,9 +53,11 @@ function get( site ) {
 			type: 'ActionStep',
-			placement: 'beside',
-			text: i18n.translate( '{{strong}}Preview:{{/strong}} Click here to see what your site looks like.', {
+			iconText: "your site's name",
```
Needs `i18n.translate()`

mcsf replied:
Also, importantly, the `translate` call needs to provide context (read: the sentence that this fragment will be inserted into). It's best that we have some certainty that the container sentence won't be changing a lot, as that may imply retranslating the fragment, even if the fragment remains unchanged in English.

lsinger commented on client/layout/guided-tours/steps.js:
```diff
@@ -18,16 +18,18 @@ class BasicStep extends Component {
+				arrow,
+				stepPos
+			} ),
+		].filter( Boolean );
```
Sweet! Thanks.

lsinger commented:
Squashed, rebased, and addressed @coreh's and @mcsf's comments wrt i18n.

ehg commented on client/layout/guided-tours/steps.js:
```diff
@@ -131,35 +145,62 @@ class ActionStep extends Component {
 		let components = {};
 		if ( this.props.icon ) {
-			components.gridicon = <Gridicon icon={ this.props.icon } size={ 24 } />
+			components.nextNudge = <span>the <Gridicon icon={ this.props.icon } size={ 24 } /></span>;
```
Needs i18n.

ehg replied:
3e63ac9

ehg commented:
LGTM. Gonna merge this :)
